client/gc.go

```diff
@@ -0,0 +1,329 @@
+}
+
+func (pq GCAllocPQImpl) Less(i, j int) bool {
+	return pq[i].timeStamp.Before(pq[j].timeStamp)
+}
+
+func (pq GCAllocPQImpl) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *GCAllocPQImpl) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*GCAlloc)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *GCAllocPQImpl) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// IndexedGCAllocPQ is an indexed PQ which maintains a list of allocation runner
+// based on their termination time.
+type IndexedGCAllocPQ struct {
+	index map[string]*GCAlloc
+	heap  GCAllocPQImpl
+
+	pqLock sync.Mutex
+}
+
+func NewIndexedGCAllocPQ() *IndexedGCAllocPQ {
+	return &IndexedGCAllocPQ{
+		index: make(map[string]*GCAlloc),
+		heap:  make(GCAllocPQImpl, 0),
+	}
+}
+
+func (i *IndexedGCAllocPQ) Push(ar *AllocRunner) error {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	alloc := ar.Alloc()
+	if _, ok := i.index[alloc.ID]; ok {
+		return fmt.Errorf("alloc %v already being tracked for GC", alloc.ID)
+	}
+	gcAlloc := &GCAlloc{
+		timeStamp:   time.Now(),
+		allocRunner: ar,
+	}
+	i.index[alloc.ID] = gcAlloc
+	heap.Push(&i.heap, gcAlloc)
+	return nil
+}
+
+func (i *IndexedGCAllocPQ) Pop() *GCAlloc {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	if len(i.heap) == 0 {
+		return nil
+	}
+
+	gcAlloc := heap.Pop(&i.heap).(*GCAlloc)
+	delete(i.index, gcAlloc.allocRunner.Alloc().ID)
+	return gcAlloc
+}
+
+func (i *IndexedGCAllocPQ) Remove(allocID string) (*GCAlloc, error) {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	if gcAlloc, ok := i.index[allocID]; ok {
+		heap.Remove(&i.heap, gcAlloc.index)
+		delete(i.index, allocID)
+		return gcAlloc, nil
+	}
+
+	return nil, fmt.Errorf("alloc %q not present", allocID)
+}
+
+func (i *IndexedGCAllocPQ) Length() int {
+	i.pqLock.Lock()
+	defer i.pqLock.Unlock()
+
+	return len(i.heap)
+}
+
+// AllocGarbageCollector garbage collects terminated allocations on a node
+type AllocGarbageCollector struct {
+	allocRunners   *IndexedGCAllocPQ
+	statsCollector stats.NodeStatsCollector
+	reservedDiskMB int
+	logger         *log.Logger
+	shutdownCh     chan struct{}
+}
+
+// NewAllocGarbageCollector returns a garbage collector for terminated
+// allocations on a node.
+func NewAllocGarbageCollector(logger *log.Logger, statsCollector stats.NodeStatsCollector, reservedDiskMB int) *AllocGarbageCollector {
+	gc := &AllocGarbageCollector{
+		allocRunners:   NewIndexedGCAllocPQ(),
+		statsCollector: statsCollector,
+		reservedDiskMB: reservedDiskMB,
+		logger:         logger,
+		shutdownCh:     make(chan struct{}),
+	}
+	go gc.run()
+
+	return gc
+}
+
+func (a *AllocGarbageCollector) run() {
+	ticker := time.NewTicker(gcInterval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := a.keepUsageBelowThreshold(); err != nil {
+				a.logger.Printf("[ERR] client: error GCing allocation: %v", err)
```

Comment:
The error could be coming in from the stats collector and not necessarily while destroying a particular allocation runner.